### PR TITLE
Consider warnings as error, fix all warnings

### PR DESCRIPTION
**Pull request by mkristien**
## Description
We want to compile without any warnings. This PR fixes all current compilation warnings and adds a compiler flag that treats warnings as error to fail future compilations with any warnings.

### CODE REVIEW

**Comment by neilweidinger** [comment]
Hmm interesting, the first commit actually didn't compile on my mac. Some files run into `-Wunused-private-field` warnings as well as others, causing the build using clang++ to fail. I tested this out on travis on osx and it fails there too, so I think we should add osx back.

Also, using `-Werror` might not be such a good idea, as expressed in [this blog post](https://embeddedartistry.com/blog/2017/05/22/werror-is-not-your-friend/). Using the flag caused the build to fail on my machine although the source code didn't change, for example. I still think we should definitely aim for zero warnings though.

Also not sure if `utils/config.mk` is the best idea either, spent a lot of time wondering where certain variables in the Makefile were defined only to remember that `utils/config.mk` exists.

File: src/utils/system.cpp [hunk 1]
```diff
@@ -205,7 +205,7 @@ System::System(int argc, char* argv[])
         else        verbose_embrakes = true;
         break;
       case 'C':
-        strncpy(config_file, optarg, 250);
+        strncpy(config_file, optarg, 250-1);
```
Comment (neilweidinger) [inline src/utils/system.cpp:208]: What's the reason for subtracting by one here?

Comment (mkristien) [reply src/utils/system.cpp:208]: This is to account for `\0` termination character in strings. Otherwise, a warning about string size is produced.

**Comment by mkristien** [comment]
I agree it might not be compiler portable. Do you have an alternative way to enforcing no-warnings policy? Last year, we talked about warnings but they still stays until this commit, i.e. they were even carried over to this year's repo. Shall we explicitly list warnings we want to be errors?

`utils/config.mk` is used to that multiple `Makefile`s can use the same compiler configurations. Sorry the build infrastructure is not obvious to users. It can be a bit messy but we are trying to make it easy to "use" not easy to "modify". It is important for everybody to easily modify the build system (e.g. change compiling flags, add new custom targets)? Maybe we shall have a workshop on using and writing `Makefile`s.

**Comment by neilweidinger** [comment]
Unfortunately I don't have a better solution for enforcing a no warning policy. Maybe qa can keep track of warnings and alert other members of them?

Ah yea I forgot about the Makefile in testing that uses `utils/config.mk`. Maybe we can move this out into the root directory, like `config.txt`? Or maybe even just make a config directory.

**Review by neilweidinger — APPROVED** [review]

**Comment by mkristien** [comment]
How about we have a list explicitly stating which warning we care about and only make compilation fail on those warnings? Hopefully, we can control compiler portability better this way. Also, everybody would need to avoid the same set of warnings.

About moving config to the root folder, wasn't this [tidy](https://github.com/Hyp-ed/hyped-2020/pull/10) PR all about making the root directory clean? In fact, I am working on a new [configuration infrastructure](https://github.com/Hyp-ed/hyped-2020/tree/config-nested) and part of it is to move `config.txt` from the root folder into a `configurations/` folder.

**Comment by neilweidinger** [comment]
Avoiding a minimum set of explicit warnings is not a bad idea. #10 was about moving things out of the root directory that don't necessarily need to be there, having `config.mk` in the root directory might be useful as that's also where the Makefile is.

**Pull request merged**